### PR TITLE
rework device sharing in volcano

### PR DESCRIPTION
**Pull request by archlitchi**
We implement a common interface for shareable devices(GPU,NPU,FPGA,...) called sharedDevicePool, and use it to reimplement current gpu-share mechanism. The goal is to let device-sharing easy to implement, and better organised. If you wish to grant vc-scheduler the ability to share another device, all you need to do is to implement these methods in sharedDevicePool, and place your logic under pkg/scheduler/api/devices. No other hackings are needed.

### CODE REVIEW

**Review by hwdef — COMMENTED** [review]
Please make CI happy first.
And squash your code.

File: pkg/scheduler/api/devices/nvidia/gpushare/device_info.go [hunk 1]
```diff
@@ -166,47 +166,48 @@ func (gs *GPUDevices) MonitorStatus() string {
 }
 
 func (gs *GPUDevices) AllocateToPod(kubeClient kubernetes.Interface, pod *v1.Pod) error {
+	klog.V(4).Infoln("DeviceSharing:Into AllocateToPod", pod)
```
Comment (Thor-wl) [inline pkg/scheduler/api/devices/nvidia/gpushare/device_info.go:169]: I'm not sure whether the parameter `pod` will be used in the log info.

Comment (archlitchi) [reply pkg/scheduler/api/devices/nvidia/gpushare/device_info.go]: thanks, you are right, now i'm using pod.Name instead

File: pkg/scheduler/api/devices/nvidia/gpushare/device_info.go [hunk 2]
```diff
@@ -0,0 +1,217 @@
+/*
+Copyright 2020 The Volcano Authors.
```
Comment (hwdef) [inline pkg/scheduler/api/devices/nvidia/gpushare/device_info.go:2]:
```suggestion
Copyright 2023 The Volcano Authors.
```

**Comment by zishen** [comment]
This rework so good, and can you make it a plugin to add a device？

File: pkg/scheduler/api/node_info.go [hunk 3]
```diff
@@ -70,8 +70,8 @@ type NodeInfo struct {
 	RevocableZone     string
 
 	// Used to store custom information
-	Others     map[string]interface{}
-	GPUDevices map[int]*GPUDevice
+	Others        map[string]interface{}
+	SharedDevices map[string]SharedDevicePool
```
Comment (william-wang) [inline pkg/scheduler/api/node_info.go:74]: Others is used for store custom resource. We can reuse it.

File: pkg/scheduler/api/node_info.go [hunk 4]
```diff
@@ -153,7 +153,7 @@ func NewNodeInfo(node *v1.Node) *NodeInfo {
 		nodeInfo.Allocatable = NewResource(node.Status.Allocatable).Add(nodeInfo.OversubscriptionResource)
 		nodeInfo.Capability = NewResource(node.Status.Capacity).Add(nodeInfo.OversubscriptionResource)
 	}
-	nodeInfo.setNodeGPUInfo(node)
+	nodeInfo.SharedDevices[GPUSharingDevice] = gpushare.NewGPUDevices(nodeInfo.Name, node)
```
Comment (william-wang) [inline pkg/scheduler/api/node_info.go:156]: Suggest to use `setNodeOthersResource` to encapsulation the `SharedDevices[GPUSharingDevice] = gpushare.NewGPUDevices(nodeInfo.Name, node)`

File: pkg/scheduler/api/node_info.go [hunk 5]
```diff
@@ -376,7 +344,7 @@ func (ni *NodeInfo) SetNode(node *v1.Node) {
 // setNode sets kubernetes node object to nodeInfo object without assertion
 func (ni *NodeInfo) setNode(node *v1.Node) {
 	ni.setOversubscription(node)
-	ni.setNodeGPUInfo(node)
+	ni.SharedDevices[GPUSharingDevice] = gpushare.NewGPUDevices(ni.Name, node)
 	ni.setRevocableZone(node)
```
Comment (william-wang) [inline pkg/scheduler/api/node_info.go:348]: same as above

File: pkg/scheduler/api/node_info.go [hunk 6]
```diff
@@ -395,13 +363,13 @@ func (ni *NodeInfo) setNode(node *v1.Node) {
 			ni.Idle.sub(ti.Resreq) // sub without assertion
 			ni.Releasing.Add(ti.Resreq)
 			ni.Used.Add(ti.Resreq)
-			ni.AddGPUResource(ti.Pod)
+			ni.SharedDevices[GPUSharingDevice].AddResource(ti.Pod)
```
Comment (william-wang) [inline pkg/scheduler/api/node_info.go:366]: suggest to retain `AddResource` and put the `SharedDevices[GPUSharingDevice].AddResource(ti.Pod)` inside `AddResource`.

File: pkg/scheduler/api/shared_device_pool.go [hunk 7]
```diff
@@ -0,0 +1,25 @@
+package api
```
Comment (william-wang) [inline pkg/scheduler/api/shared_device_pool.go:1]: please add the copyright and change the name of this file to device_interface.go

File: pkg/scheduler/api/shared_device_pool.go [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+import (
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	GPUSharingDevice = "GpuShare"
+)
+
+type SharedDevicePool interface {
```
Comment (william-wang) [inline pkg/scheduler/api/shared_device_pool.go:12]: The interface covers the gpu number as well and npu in the future. It's better to change the name to from `SharedDevicePool` to `Devices`.

File: pkg/scheduler/api/shared_device_pool.go [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+	//following two functions used in node_info
+	AddResource(pod *v1.Pod)
```
Comment (william-wang) [inline pkg/scheduler/api/shared_device_pool.go:14]: please add function description for each interface.

File: pkg/scheduler/api/shared_device_pool.go [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+	SubResource(pod *v1.Pod)
+
+	//following four functions used in predicate
+	RequestInPod(pod *v1.Pod) bool
```
Comment (william-wang) [inline pkg/scheduler/api/shared_device_pool.go:18]: suggest to change `RequestInPod` to `HasDeviceRequest` or some other name that is easy to understand.

File: pkg/scheduler/api/shared_device_pool.go [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+	FitInPod(pod *v1.Pod) (bool, error)
```
Comment (william-wang) [inline pkg/scheduler/api/shared_device_pool.go:19]: suggest to change interface from `FitInPod` to `Filter` or `FilterNode`

File: pkg/scheduler/api/shared_device_pool.go [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+	AllocateToPod(kubeClient kubernetes.Interface, pod *v1.Pod) error
```
Comment (william-wang) [inline pkg/scheduler/api/shared_device_pool.go:20]: change `AllocateToPod` and `ReleaseFromPod` to `Allocate` and `Release`

File: pkg/scheduler/api/shared_device_pool.go [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+	ReleaseFromPod(kubeClient kubernetes.Interface, pod *v1.Pod) error
+
+	//used for debug and monitor
+	MonitorStatus() string
```
Comment (william-wang) [inline pkg/scheduler/api/shared_device_pool.go:24]: change `MonitorStatus` to `GetStatus` or `CheckStatus`

**Comment by archlitchi** [comment]
> This rework so good, and can you make it a plugin to add a device？

I'd like to, but you still need to hack into scheduler.api even with this PR merged (ie. you must put your device-initialization logic in node_info.go in order for the scheduler to recognize your device). so it's hard for a plugin to do that

**Pull request merged**